doc/compiler/il/MainlineAndOutOfLineCode.md

```diff
@@ -0,0 +1,47 @@
+<!--
+Copyright (c) 2021, 2021 IBM Corp. and others
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Mainline and Out of Line Code
+The operations that are believed to run commonly are arranged on the mainline path. The less often executed operations are chosen as the branch target on the out of line (`OOL`) path. Arranging the code that executes more often in the mainline takes advantage of branch prediction and instruction cache resources. The out of line path is usually in a different cache line for an infrequently executed operation. By falling through on the mainline path, the work could be done in the same cache line, which avoids jumping to another spot to bring in an entire new cache line to the cache. Occasionally, there would be some minor branching within the code on the mainline path to implement some basic logics.
```

Comment:
> By falling through on the mainline path, the work could be done in the same cache line, which avoids jumping to another spot to bring in an entire new cache line to the cache.

This isn't very clear to me. How about:

By introducing locality to the most frequently executed parts of code and outlining the least frequently executed fallback paths, we increase the probability of reducing the number of cache lines needed to execute a particular code sequence.

Reply:
If you wanted to elaborate further on branch prediction benefits, we can add something like:

This feature works more optimally with hardware branch prediction, as 1) predictions are generally optimized for the fall-through path by many hardware platforms, and 2) predicted fall-through instructions are generally already loaded in the instruction cache (i-cache), thereby avoiding potential branch prediction delays due to i-cache misses for predicted _taken_ paths.

Reply:
Good suggestions @fjeremic @joransiu! Thanks! I'll add them in the next update.

Reply:
Added in 6738cf6